### PR TITLE
Add hardhat deployment and test

### PR DESCRIPTION
**Pull request by boyuan-chen**
- Add hardhat deployment scripts
- Add CI/CD using hardhat deployment scripts to boot up the devnet and run the tests

To start the hardhat deployment, run `make devnet-hardhat-up`. To test the devnet, run `make devnet-hardhat-test`.

Note:
The `genesis-l2.json` and `rollup.json` are created via `boba-chain-ops`.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -1009,6 +1009,93 @@ jobs:
...
+      - run:
+          name: Install and build
+          command: |
+            pnpm install --frozen-lockfile --prefer-offline && pnpm build
+      - run:
+          name: generate cannon prestate
+          command: make cannon-prestate
+      - run:
+          name: Build images
+          command: docker-compose build
```
Comment (jyellick) [inline .circleci/config.yml]: Should any of these steps be coupled to the make target?  I realize much of the above is just environment setup for CI, but anytime I see such big CI definitions, it makes me worry about how easy it is for developers to replicate a failing CI step.

**Comment by jyellick** [comment]
Tried running `make devnet-hardhat-up` locally -- the first time it hung and failed to complete.  The second time I was able to get it to succeed.

I'm seeing a tracked file change afterwards though:

```
packages/contracts-bedrock/deploy-config/hardhat-local.json
```

Once it was up, `make devnet-hardhat-test` worked.

**Comment by boyuan-chen** [comment]
> Tried running `make devnet-hardhat-up` locally -- the first time it hung and failed to complete. The second time I was able to get it to succeed.
> 
> I'm seeing a tracked file change afterwards though:
> 
> ```
> packages/contracts-bedrock/deploy-config/hardhat-local.json
> ```
> 
> Once it was up, `make devnet-hardhat-test` worked.

Yes, the timestamp and l1 hash in the configuration file will change after deploying the contracts.

**Comment by jyellick** [comment]
> Yes, the timestamp and l1 hash in the configuration file will change after deploying the contracts.

Should we be tracking this file then?  This was a complaint about the existing `devnet-up` that we were fixing upstream, so it would be great if we could figure out a way to ignore these changes.

**Comment by boyuan-chen** [comment]
> > Yes, the timestamp and l1 hash in the configuration file will change after deploying the contracts.
> 
> Should we be tracking this file then? This was a complaint about the existing `devnet-up` that we were fixing upstream, so it would be great if we could figure out a way to ignore these changes.

I can restore the file after booting up the system. We don't need it after generating `rollup.json` and `genesis-l2.json` files.

**Comment by boyuan-chen** [comment]
@jyellick I added the code to restore the configuration file after the deployment. The changes will not exist.

**Review by jyellick — APPROVED** [review]
I'm able to run the make targets successfully and observe no changes to tracked files.

**Review by mmontour1306 — APPROVED** [review]

File: .circleci/config.yml (hunk 1 above)
Comment (boyuan-chen) [reply .circleci/config.yml]: The `make cannon-prestate` is already included in `make devnet-hardhat-up`. The `docker-compose build` is not necessary to be invoked before the `make devnet-hardhat-up`. Both commands are removed.

**Comment by boyuan-chen** [comment]
The pre-deployed BOBA token test is added.

The process will deploy a new L1 BOBA token and bridge it between L1 and L2 in the test. The L2 BOBA token is pre-deployed at `0x4200000000000000000000000000000000000023`.

**Pull request merged**